snuba/datasets/tags_column_processor.py

```diff
@@ -157,93 +152,59 @@ def __tag_expr(self, parsed_col: ParsedNestedColumn, table_alias: str = "",) ->
-                literals = condition.parameters[1].parameters
-                return [
-                    str(literal.value)
-                    for literal in literals
-                    if isinstance(literal, Literal)
-                ]
-
-        if is_binary_condition(condition, BooleanFunctions.AND):
-            assert isinstance(condition, FunctionCall)
-            return self.__extract_top_level_tag_conditions(
-                condition.parameters[0]
-            ) + self.__extract_top_level_tag_conditions(condition.parameters[1])
-
-        return []
-
     def __get_filter_tags(self, query: Query) -> List[str]:
         """
         Identifies the tag names we can apply the arrayFilter optimization on.
         Which means: if the tags_key column is in the select clause and there are
         one or more top level conditions on the tags_key column.
+
+        We can only apply the arrayFilter optimization to tag keys conditions
+        that are not in OR with other columns. To simplify the problem, we only
+        consider those conditions that are included in the first level of the query:
+        [['tagskey' '=' 'a'],['col' '=' 'b'],['col2' '=' 'c']]  works
+        [[['tagskey' '=' 'a'], ['col2' '=' 'b']], ['tagskey' '=' 'c']] does not
         """
-        if not state.get_config("ast_tag_processor_enabled", 0):
+        if not state.get_config("ast_tag_processor_enabled", 1):
             return []
 
-        select_clause = query.get_selected_columns_from_ast() or []
-
         tags_key_found = any(
-            f.function_name == "arrayJoin"
-            and f.parameters == (Column(None, None, "tags.key"),)
-            for expression in select_clause
-            for f in expression
-            if isinstance(f, FunctionCall)
+            "tags_key" in columns_in_expr(expression)
+            for expression in query.get_selected_columns() or []
         )
 
         if not tags_key_found:
             return []
 
         def extract_tags_from_condition(
-            cond: Optional[Expression],
+            cond: Sequence[Condition],
         ) -> Optional[List[str]]:
             if not cond:
                 return []
-            if any(is_binary_condition(cond, BooleanFunctions.OR) for cond in cond):
-                return None
-            return self.__extract_top_level_tag_conditions(cond)
 
-        cond_tags_key = extract_tags_from_condition(query.get_condition_from_ast())
+            ret = []
+            for c in cond:
+                if not is_condition(c):
+                    # This is an OR
+                    return None
+
+                if c[1] == "=" and c[0] == "tags_key" and isinstance(c[2], str):
+                    ret.append(c[2])
+
+                elif (
+                    c[1] == "IN"
+                    and c[0] == "tags_key"
+                    and isinstance(c[2], (list, tuple))
+                ):
+                    ret.extend([str(tag) for tag in c[2]])
```

Comment:
How come we only cast the value to string here when our operator is "IN" and not "=" on line 191?

Reply:
Technically it should not be needed in either case, but I don't trust the content of the legacy query representation, since it can be modified in fairly unpredictable ways. Will add it there as well.